File: problem-solving/lc-1143-medium-longest-common-subsequence/1-top-down.go
/*
https://leetcode.com/problems/longest-common-subsequence/

Complexity
----------
Time: O(m*n)
Space: O(m*n)
*/

func longestCommonSubsequence(text1 string, text2 string) int {
    cache := make(map[string]int)
    var F func(int, int) int
    F = func(i, j int) int {
        if i == len(text1) || j == len(text2) {
            return 0
        }
        key := fmt.Sprintf("%d-%d", i, j)
        if v, ok := cache[key]; ok {
            return v
        }
        if text1[i] == text2[j] {
            cache[key] = 1 + F(i + 1, j + 1)
        } else {
            cache[key] = max(F(i + 1, j), F(i, j + 1))
        }
        return cache[key]
    }
    return F(0, 0)
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

File: problem-solving/lc-120-medium-triangle/1-top-down.py
"""
https://leetcode.com/problems/triangle/

Complexity
----------
Time: O(n*n)
Space: O(n*n)
"""

class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        n = len(triangle)
        cache = {}
        def solve(r, c):
            if r >= n:
                return 0
            if (r, c) in cache:
                return cache[(r, c)]
            cache[(r, c)] = triangle[r][c] + min(solve(r + 1, c),
                                                 solve(r + 1, c + 1))
            return cache[(r, c)]
        return solve(0, 0)
